alternatives.go

```
@@ -0,0 +1,105 @@
+	"os/exec"
+	"path/filepath"
+)
+
+// iptablesAlternativeSelector allows to configure a system to use iptables in
+// nft or legacy mode.
+type iptablesAlternativeSelector interface {
+	// useMode configures the system to use the selected iptables mode.
+	useMode(ctx context.Context, mode iptablesMode) error
+}
+
+// buildIptablesAlternativeSelector builds the proper iptablesAlternativeSelector depending
+// on the machine's setup. It will use either `alternatives` or `update-alternatives` if present
+// in the sbin folder. If none is present, it will manage iptables binaries by manually
+// creating symlinks.
+func buildIptablesAlternativeSelector(sbinPath string) iptablesAlternativeSelector {
+	if executableExists(filepath.Join(sbinPath, "alternatives")) {
+		return alternativesSelector{sbinPath: sbinPath}
+	} else if executableExists(filepath.Join(sbinPath, "update-alternatives")) {
+		return updateAlternativesSelector{sbinPath: sbinPath}
+	} else {
+		// if we don't find any tool to managed the alternatives, handle it manually with symlinks
+		return symlinkSelector{sbinPath: sbinPath}
+	}
+}
+
+// updateAlternativesSelector manages an iptables setup by using the `update-alternatives` binary.
+// This is most common for debian based OSs.
+type updateAlternativesSelector struct {
+	sbinPath string
+}
+
+func (u updateAlternativesSelector) useMode(ctx context.Context, mode iptablesMode) error {
+	modeStr := mode.String()
+	var errors []error
+
+	if err := runAndReadError(exec.CommandContext(ctx, "update-alternatives", "--set", "iptables", "/usr/sbin/iptables-"+modeStr)); err != nil {
+		errors = append(errors, fmt.Errorf("update-alternatives iptables to mode %s: %v", modeStr, err))
+	}
+
+	if err := runAndReadError(exec.CommandContext(ctx, "update-alternatives", "--set", "ip6tables", "/usr/sbin/ip6tables-"+modeStr)); err != nil {
+		errors = append(errors, fmt.Errorf("update-alternatives ip6tables to mode %s: %v", modeStr, err))
+	}
+
+	return joinErrors(errors)
```

If the first command fails then you can just bail out immediately (since the second will almost certainly also fail in exactly the same way, and even if it doesn't, the user is still screwed).

You could also simplify by just using `logErrorAndExit` rather than returning an error.

---

I only did this to follow more closely (I think) what the current script does today

I prefer to return the error and exit from main though, I think makes the program easier to read.

---

FWIW, the current script does it this way because that's what was simplest when writing it as a script. There's no deeper reason for it, and no reason to copy that behavior if it's _not_ simpler in Go form (eg, needing to have `joinErrors`).

---

Got you, removed. I did the same for the symlink alternative as well.